docs/phdata.rst

```diff
@@ -397,6 +407,60 @@ The following fields are also arrays:
   This field is not relevant when no polarization- and spectral-insensitive
   splitting is performed.
 
+The following fields can be used to indicate the meaning of space-time markers.
+
+- **num_space_time_markers**: (integer) *New in version 0.6.* number of
+  space-time markers. Required if space-time markers are recorded. This value
+  must match the number of space-time markers specified by
+  ``detectors_specs/space_time_marker1``,
+  ``detectors_specs/space_time_marker2``, and so on.
+
+- **space_time_markers**: (array of strings) *New in version 0.6.* for each
+  space-time marker, this field indicates a standard interpretation. The length
+  of this array must equal the value of ``num_space_time_markers``. Currently
+  defined interpretations are for indicating raster timings in laser-scanning
+  imaging (such as FLIM), and valid element values are:
+
+  - ``"pixel"`` (pixel clock; the marker indicates the beginning timestamp of
+    each raster pixel),
+  - ``"line"`` (line clock; the marker indicates the beginning timestamp of each
+    raster scanline),
+  - ``"frame"`` (frame clock; the marker indicates the beginning timestamp of
+    each raster frame), and
+  - ``""`` (empty string; indicates a marker not used for any of the above
+    standard meanings).
+
+.. _raster_group:
+
+Raster group
+^^^^^^^^^^^^
+
+*New in version 0.6.*
+
+The optional **/setup/raster** group contains information about raster scan
+measurements (typically FLIM, but could be any single-photon-counting
+raster-scan imaging method).
+
+These 3 fields are required if this group is present:
```

Comment:
Do we want to add a `raster_type` string under the `setup/raster` group so that we can leave open other non-raster scanning spatial modulations, each with its own set of maker to space-time conversion definitions?

We should specify if the `raster_type` is not `raster` then and additional field, maybe `pixel_coordinates` needs to be specified where the location of each pixel is located...

Reply:
Perhaps `space_time_markers_type` or something, so that it's not called "raster" when it's not "raster"?

It does seem like a good idea to have this, even if we only add one type initially, for future extensibility.

(I don't have enough experience with non-raster scanning to know if recording _a priori_ pixel coordinates is always possible or desirable (or even sufficient to recover an image). It may make sense for small scans, but I would expect for large scans one would want to record parameters for the curves instead.)

One less ambitious type that we might consider is bidirectional raster scans (in which alternating lines are scanned in opposite directions), though I don't feel this is important enough to put into an initial version.

Reply:
And yet, bi-directional is not as simple as it seems. The frame and pixel clock (if present) need not change, but the line clock (if present) could be a single clock every line or every other line, or two separate clocks for each direction. Also, at least in theory the first line could be left-to-right or right-to-left. And if two separate clocks are used, the direction of the first line could vary from frame to frame (again, at least in theory).

But this might be a good illustration of the situation we face: if we try to create a spec that covers all these possibilities, it will be complicated and hard to understand (and use correctly), it may include features that are theoretical and never end up being used, and after all that, we could easily miss some possibility (or a new variant may be invented in the future).

I think it is much more important to design for incremental changes, so that each space-time-markers-type (if we decide to go with this way) is as simple as possible, and new ones can be added with less work. Perhaps this means having more than one type for bi-directional (rectangular) rasters to account for the different patterns. That way, over time, we can add the patterns that are actually encountered (or proposed) in hardware.

In the bi-directional scanning case, for example, we could add a markers-type for the case of a single line clock recorded for every line. Support for any other pattern can later be added if/when we actually find the need, either as a a different markers-type or by updating the spec for the markers-type-specific fields.

(Personally I think it is better to get out a new release of the spec with the minimal and by far most common case of uniform, unidirectional, rectangular raster scans, leaving the addition of support for the other possibilities for subsequent releases.)